### PR TITLE
Updated readthedocs.yml to linkcheck

### LINKED ISSUE #3387: Add a link checker step in the doctests jobs and fix broken links that appear
Currently, we use Lychee to check for broken links, but we ignore a lot of Sphinx-related links because of referencing files internally. `Sphinx` provides a [`linkcheck`](https://www.sphinx-doc.org/en/master/usage/builders/index.html#sphinx.builders.linkcheck.CheckExternalLinksBuilder) builder for this purpose that can be invoked in CI for checking links (either through a new `nox` session, or within the `docs` session using posargs, or just calling the command directly).

It might make sense to rename the Lychee workflow to something like `links.yml` and add a parallel job to run the above command instead of another step.

### PR DESCRIPTION
**Pull request by RohitP2005**
# Description

This PR introduces a new pre-build step to the `readthedocs.yml` configuration for PyBaMM. The change adds a `linkcheck` operation to ensure that all links in the documentation are checked for validity during the build process on Read the Docs.

### Changes made:
- Added a `pre_build` step to the `readthedocs.yml` file.
- The `pre_build` step runs the `sphinx -b linkcheck` command, which checks all links in the documentation before the build proceeds.
- The command is configured to have a short timeout (`linkcheck_timeout=1`) to handle any slow links more efficiently.

This ensures that if any broken links exist in the documentation, they are caught early during the build process.

**Fixes #3387**

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python -m pytest` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python -m pytest --doctest-plus src` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by kratman** [comment]
Previous discussion: #4686

**Comment by agriyakhetarpal** [comment]
Thanks, this looks reasonable. As mentioned, you can ignore some of the links that Lychee is ignoring as well, but the rest of the failing links will either have to be removed, replaced, or fixed before we can approve this.

**Comment by RohitP2005** [comment]
@agriyakhetarpal understood, in the next commit I will add the links to be ignored and then we can work on the failing links

**Comment by RohitP2005** [comment]
I ran the build locally and have logs of the broken links. I’ve added the ScienceDirect links to be ignored. For the other broken links, we can ignore the warnings temporarily to let the build complete and address them later. Let me know if you'd prefer to prioritize fixing them now.

File: .readthedocs.yaml
Comment (agriyakhetarpal) [inline .readthedocs.yaml:22]: Could you please explain what this is for? It looks like it breaks the build: 
https://readthedocs.org/projects/pybamm/builds/26675902/

**Review by agriyakhetarpal — COMMENTED** [review]
Yes, my preference is that we fix all of the broken links, since based on the previous RTD build, there are not many (around 13): https://readthedocs.org/projects/pybamm/builds/26675630/

You may refer to https://www.sphinx-doc.org/en/master/usage/referencing.html#role-ref which describes the proper way to reference headings in Sphinx. That is, you need to add labels for the headings that are being reported as broken that can be used as a reference (and reference the labels if they already exist).

For example, for the broken link for the IREE solver (line 42 in `installation/index.rst`), we already have a reference in line 108 of `installation/gnu-linux-mac.rst` as follows:

```rest
.. _optional-iree-mlir-support:
```

Hence, this change should fix it:

```patch
- *  `IREE <https://iree.dev/>`_ (`MLIR <https://mlir.llvm.org/>`_) support, see `Optional - IREE / MLIR Support <gnu-linux-mac.html#optional-iree-mlir-support>`_.
+ *  `IREE <https://iree.dev/>`_ (`MLIR <https://mlir.llvm.org/>`_) support, see :ref:`optional-iree-mlir-support`.
```

and similarly for the other internal broken links. For external broken links, if there is an easy replacement, we should do that, otherwise we should add them to the ignore list too..

File: .readthedocs.yaml
Comment (RohitP2005) [reply .readthedocs.yaml:22]: This variable is  intended to define the output directory for build artifacts like link checks or PDFs.

**Comment by RohitP2005** [comment]
## Refernce for broken links
### Internal Broken Links
1. `gnu-linux-mac.html#optional-iree-mlir-support`
2. `gnu-linux-mac.html#optional-jaxsolver`
3. `index.html#dependencies`
4. `index.html#install-required-dependencies`
5. `index.html#installation`
6. `install-from-source.html`
7. `source/user_guide/contributing.html`
8. `source/user_guide/installation/index.html`
9. `windows-wsl.html`

### External Broken Links
1. `https://books.google.co.uk/.../id=QazcnD7GUoUC...`
2. `https://doi.org/10.1137/20M1336898`
3. `https://en.wikipedia.org/wiki/Rectifier_(neural_networks`
4. `https://github.com/pybamm-team/.../CONTRIBUTING.md#a-before-you-begin`

**Comment by RohitP2005** [comment]
The build ran successfully locally , but failed @agriyakhetarpal 
image

**Comment by agriyakhetarpal** [comment]
Yes, there are some error messages (please see the Read the Docs job) such as this: 

`/home/docs/checkouts/readthedocs.org/user_builds/pybamm/checkouts/4712/docs/source/user_guide/installation/index.rst:153: CRITICAL: Title level inconsistent:`

which now need to be fixed. You may refer to https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections to see how Sphinx processes different levels of headings, which might be helpful.

Also, I'm not sure `linkcheck_timeout=1` might be too less to be able to check GitHub URLs.

Thanks for staying put on this, it's almost there!

**Comment by RohitP2005** [comment]
@agriyakhetarpal  I will look into it.
I am also looking forward to complete this task.

**Comment by RohitP2005** [comment]
# Sphinx Build Error: Title Level Inconsistency in `index.rst`

## Errors:

The following titles in the `index.rst` file have inconsistent title levels:

1. **Plot dependencies**  
   Line 98: `Plot dependencies` title has inconsistent level.

2. **Docs dependencies**  
   Line 112: `Docs dependencies` title has inconsistent level.

3. **Examples dependencies**  
   Line 140: `Examples dependencies` title has inconsistent level.

4. **Dev dependencies**  
   Line 153: `Dev dependencies` title has inconsistent level.

5. **Cite dependencies**  
   Line 176: `Cite dependencies` title has inconsistent level.

6. **bpx dependencies**  
   Line 189: `bpx dependencies` title has inconsistent level.

7. **tqdm dependencies**  
   Line 202: `tqdm dependencies` title has inconsistent level.

8. **Jax dependencies**  
   Line 215: `Jax dependencies` title has inconsistent level.

9. **IREE dependencies**  
   Line 227: `IREE dependencies` title has inconsistent level.

10. **Full installation guide**  
    Line 238: `Full installation guide` title has inconsistent level.

**Comment by RohitP2005** [comment]
The build is still failing , i will refactor the headings and make another commit

**Comment by RohitP2005** [comment]
@agriyakhetarpal can u help me, why the build is still failing?
 cause i even fixed the warnings also

**Comment by RohitP2005** [comment]
The timeout fixed it i suppose

File: docs/conf.py
Comment (kratman) [inline docs/conf.py:93]: Why is this being ignored? It seems weird that we cannot check a link to our own code base. I would rather have a different workaround than ignore this link

File: docs/source/user_guide/installation/index.rst
Comment (kratman) [inline docs/source/user_guide/installation/index.rst:7]: This seems out of place. This is the main installation page for PyBaMM, not a page for required dependencies

**Review by kratman — CHANGES_REQUESTED** [review]
Partial review

**Comment by kratman** [comment]
The dropdown menus on the docs seem to have been broken:

Before:
image

Now:
image

File: docs/source/user_guide/installation/index.rst
Comment (kratman) [inline docs/source/user_guide/installation/index.rst:89]: This is needed please put it back

**Review by kratman — CHANGES_REQUESTED** [review]
This is the cause of the missing dropdown menus for subsections. Please put these back. This PR should only be updating URLs not restructuring the documentation

File: docs/conf.py
Comment (agriyakhetarpal) [reply docs/conf.py:93]: I think the reason for this was that GitHub URLs are never happy with link checkers and have aggressive rate limits (especially because we can't supply `GITHUB_TOKEN` unlike how we can Lychee)

IMO, it seems fine to ignore because Lychee is checking this URL anyway, but we should add a comment about it somewhere.

Comment (kratman) [reply docs/conf.py:93]: I still think it is better to find another workaround. You can add secrets to read the docs if that is the problem with an alternative fix

File: docs/source/user_guide/installation/index.rst
Comment (RohitP2005) [reply docs/source/user_guide/installation/index.rst:7]: sorry about that

File: docs/source/user_guide/installation/index.rst
Comment (RohitP2005) [reply docs/source/user_guide/installation/index.rst:89]: will tend to it

File: docs/conf.py
Comment (agriyakhetarpal) [reply docs/conf.py:93]: No, my point is that it might be better that the Sphinx linkcheck can be reserved to docs-only links, and Lychee can check the rest of the links, which was my idea when I opened #3387.

Adding secrets to the Read the Docs config sounds like another nightmare when everything is moving towards environments and trusted publishing, especially for a GitHub App like it that's already hogging up many permissions when setting it up. So, I wouldn't really recommend that.

At this moment, the Contributing guide is just `..include ::`d into the docs, but it resides outside it in the repository's root (I implemented this change back then), so one reasonable workaround is to replace the GitHub link with the RTD link:

`https://docs.pybamm.org/en/latest/source/user_guide/contributing.html#a-before-you-begin` (note the "latest" string in the URL because it's a development-related document).

File: docs/source/user_guide/installation/index.rst
Comment (kratman) [reply docs/source/user_guide/installation/index.rst]: To clarify, this should be put back for all sections were it changed. Not just this one line

File: docs/conf.py
Comment (kratman) [reply docs/conf.py:93]: I like the RTD link idea if it works for both

File: docs/source/user_guide/installation/index.rst
Comment (RohitP2005) [reply docs/source/user_guide/installation/index.rst]: Actually they cause inconsistency in header and made the build fail. 
I will look into it .

Comment (kratman) [reply docs/source/user_guide/installation/index.rst]: They worked before, is this a linkcheck issue? It is a basic part of restructured text

File: docs/conf.py
Comment (kratman) [reply docs/conf.py:93]: @RohitP2005 Can you do the RTD link change, that is pretty much the last thing I want from the changes

File: docs/source/user_guide/installation/index.rst
Comment (RohitP2005) [reply docs/source/user_guide/installation/index.rst]: yes, but i resolved your request with my latest commit. sry for the trouble
image

File: docs/source/user_guide/installation/index.rst
Comment (agriyakhetarpal) [inline docs/source/user_guide/installation/index.rst:48]:
```suggestion
PyBaMM requires the following dependencies:
```

File: docs/source/user_guide/installation/windows.rst
Comment (agriyakhetarpal) [inline docs/source/user_guide/installation/windows.rst:113]:
```suggestion
instructions on the `Installation <windows-wsl.rst>`_ page.
```

I think the word "instructions" was cut out when you updated the link

**Review by agriyakhetarpal — COMMENTED** [review]
I'm happy to approve once CI passes, thanks!

**Comment by kratman** [comment]
@agriyakhetarpal I just have the one last RTD link thing so we can skip the suppression

**Comment by kratman** [comment]
@all-contributors please add @RohitP2005 for docs and infra

**Comment by kratman** [comment]
@RohitP2005 We just need you to change

https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#a-before-you-begin

to

https://docs.pybamm.org/en/latest/source/user_guide/contributing.html#a-before-you-begin

and remove the link suppression. Then we can get this merged

**Review by kratman — APPROVED** [review]
Thanks

**Comment by kratman** [comment]
Just waiting for tests to pass

**Comment by RohitP2005** [comment]
> Thanks

Anytime Erick , Happy to contribute.

**Comment by RohitP2005** [comment]
@agriyakhetarpal Now that this issue is sorted , ig i will work on the other ones aswell

**Pull request merged**
